Query single repo

fifikim opened the pull request:
This Draft PR corresponds to the ticket for [Query GitHub API for a Single Repo](https://trello.com/c/y8hPwWiB)

It adds:
- uses Apollo-Angular to query the GitHub API for a single repository
- if found, full repository details are displayed
- if not found, a not found message will be displayed
- README has been updated with instructions for entering an API personal access token 
- all tests are passing

fifikim committed:
add GitHub API query and mapper service

fifikim committed:
add localhost environment to store API personal access token

fifikim committed:
update README with expanded directions for creating personal access token

fifikim committed:
update repo-detail template to use async tag for status variable

davetorre commented on src/app/mapper.service.spec.ts:
```diff
@@ -0,0 +1,41 @@
+    const actualMapped = service.mapRepos([REPO_DATA1, REPO_DATA2]);
+
+    expect(actualMapped).toEqual([MAPPED_REPO1, MAPPED_REPO2]);
+  });
```
The tests on lines 18 and 30 both failed for me because the actual datetime values were an hour before the expected. Probably because I'm in central time and you're in eastern and the expected values are hard-coded?

fifikim replied:
Ah, didn't think of that... will update the mock data to be a Date object instead of a string. Thanks!

fifikim replied:
I think I fixed this by setting Jest's timezone to UTC but will you please lmk if it's still failing on your end?

davetorre replied:
Fixed it for me 👍

fifikim committed:
update mock server data and service tests

fifikim committed:
update angular environment configurations

fifikim committed:
update build script in CI workflow

fifikim committed:
set Jest timezone to UTC

fifikim committed:
update CI workflow build script

seoulection commented on .github/workflows/CI.yml:
```diff
@@ -24,7 +24,7 @@ jobs:
       
       - name: Build
-        run: npm run build 
+        run: npm run build:prod
```
Just out of curiosity, what prompted this change?

fifikim replied:
It was necessary to get CI to pass because otherwise it uses the default `development` environment and looks for an `environment.development.ts` file which has been gitignored due to the fact that it stores the API access token

seoulection commented on package.json:
```diff
@@ -5,8 +5,8 @@
-    "test": "jest",
+    "build:prod": "ng build --configuration production",
+    "test": "TZ=utc jest",
     "test:watch": "jest --watchAll",
     "test:coverage": "jest --coverage",
```
You would also need to add `TZ=utc` here

seoulection reviewed (commented):
I'm unable to run the server 👀 

<img width="1218" alt="Screen Shot 2022-09-27 at 2 24 37 PM" src="https://user-images.githubusercontent.com/6445410/192638510-1fc4f20e-f9e0-4cbf-b3dc-7d14e96f17b4.png">

fifikim commented:
> I'm unable to run the server 👀
> 
> <img alt="Screen Shot 2022-09-27 at 2 24 37 PM" width="1218" src="https://user-images.githubusercontent.com/6445410/192638510-1fc4f20e-f9e0-4cbf-b3dc-7d14e96f17b4.png">

Did you catch the instructions I added to the README file about updating the file called `environment.development.ts sample`? 

If you had already made a `localhost` environment you'll need to rename it to `development` (sorry -- I changed the configurations when I realized that CI was failing)

fifikim committed:
update test scripts

davetorre commented:
Are you seeing these errors for one of the repo.service.spec.ts tests?
console.error
      Missing field 'refs' while writing result {
        "repository": {
          "__typename": "Repository",
          "name": "zagaku",
          "description": "Learning whilst seated",
          "url": "https://github.com/8thlight/zagaku",
...

fifikim commented:
> Are you seeing these errors for one of the repo.service.spec.ts tests? console.error Missing field 'refs' while writing result { "repository": { "__typename": "Repository", "name": "zagaku", "description": "Learning whilst seated", "url": "https://github.com/8thlight/zagaku", ...

Yes I am! I sent a message about this to the channel yesterday... spent a while trying to sort this out and couldn't figure it out. Will take another look now to see if fresh eyes help

seoulection commented on README.md:
```diff
@@ -20,6 +20,18 @@ Technologies used: Angular 14/TypeScript, RxJs, Flex-Layout, Apollo-Angular for
+- a <a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token">GitHub GraphQL API personal access token</a> linked to a profile belonging to the 8th Light GitHub organization 
+
+When creating your personal access token, request the following scopes:
+repo
+read:packages
+read:org
+read:public_key
+read:repo_hook
+user
+read:discussion
+read:enterprise
+read:gpg_key
```
This block isn't being formatted in the README, so it all is squished on one line

fifikim replied:
fixed, thx!

seoulection commented on README.md:
```diff
@@ -20,6 +20,18 @@ Technologies used: Angular 14/TypeScript, RxJs, Flex-Layout, Apollo-Angular for
 
 - <a href="https://angular.io/guide/setup-local#install-the-angular-cli">Angular 14</a>
 - <a href="https://nodejs.org/">Node.js 16.17.0</a> 
+- a <a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token">GitHub GraphQL API personal access token</a> linked to a profile belonging to the 8th Light GitHub organization 
```
Extra `a` here

seoulection commented on src/environments/environment.development.ts sample:
```diff
@@ -0,0 +1,18 @@
+// This file can be replaced during build by using the `fileReplacements` array.
```
Would we just be able to remove this `suffix` and add this file to `.gitignore` after it's been committed? Right now, changes to this file are being tracked which could allow someone to accidentally push secrets

fifikim replied:
Hmm.. I always thought this was the safe way to provide an env file template to share with others. Would it be safer if I gave explicit instructions in the README to rename the file before it's edited so that the changes are never tracked?

seoulection replied:
You'll still be able to provide the template since it's committed to the repo. I think that renaming this file to what it needs to be saves the user an extra step since they're going to have to rename it anyways. Let me know if you wanna hop on a call to discuss more!

seoulection commented on src/app/repo.service.ts:
```diff
@@ -1,22 +1,40 @@
-        repos.find(repo => repo.name === name))
-    )
+  
+  getRepo(name: any): Observable<Repository> {
```
Since the `GET_REPO` expects that `name` is a `string`, would we be able to type the `name` parameter as `string`?

fifikim replied:
Yes! updated thanks

seoulection commented on src/app/mapper.service.ts:
```diff
@@ -0,0 +1,91 @@
+export class MapperService {
+  constructor() { }
+
+  mapRepos(repos: any[]): Repository[] {
```
Non-blocking and more for future reference - generally, using `any` as a type is discouraged since it bypasses type checking (which sort of defeats the whole purpose of using Typescript). It looks like you may be able to install this library to add types to fields in API responses: https://www.npmjs.com/package/@octokit/types

fifikim replied:
Ah cool, didn't even occur to me that a type would still be expected for such a complex object. I think that library only works for REST APIs but I found an equivalent version that works for GraphQL and instructions for using it with the GitHub GraphQL API: [https://benlimmer.com/2020/05/16/adding-typescript-types-github-graphql-api/](https://benlimmer.com/2020/05/16/adding-typescript-types-github-graphql-api/)

The setup looks really complicated, though -- so since this is non-blocking I think I'll skip unless you ask me to add it!

seoulection replied:
Ah you're right, that was just for REST API endpoints 😬 but yeah anyways, just something to keep in mind if you stumble across having to use `any` as a type!

seoulection reviewed (commented):
Was able to run through the README and get things running locally. Everything looks good! I did notice that there's an invalid date for `Most Recent Pull Request` when you open a specific repo's detail page.

![Screen Shot 2022-09-28 at 7 23 31 PM](https://user-images.githubusercontent.com/6445410/192924558-55c8bc71-9480-43d6-81d2-346e7d51e5e3.png)

fifikim commented:
> Are you seeing these errors for one of the repo.service.spec.ts tests? console.error Missing field 'refs' while writing result { "repository": { "__typename": "Repository", "name": "zagaku", "description": "Learning whilst seated", "url": "https://github.com/8thlight/zagaku", ...

Yeah, I still have no idea how to clear these errors. I tried making fields nullable when writing to the InMemoryCache, and also tried disabling caching for the Apollo tests, but couldn't prevent the logs :/

> Was able to run through the README and get things running locally. Everything looks good! I did notice that there's an invalid date for `Most Recent Pull Request` when you open a specific repo's detail page.
> 
> ![Screen Shot 2022-09-28 at 7 23 31 PM](https://user-images.githubusercontent.com/6445410/192924558-55c8bc71-9480-43d6-81d2-346e7d51e5e3.png)

updated the date mapper! thanks for catching

fifikim committed:
update README and pull request mapper

seoulection reviewed (approved):
Thanks for addressing the issue with the date! I'll just approve for now since I realized the rest of my comments are sorta non-blockers
